contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol

```diff
@@ -0,0 +1,224 @@
+     * @return magicValue The magic value.
+     */
+    function isValidSignature(
+        bytes32 dataHash,
+        bytes memory moduleSignature
+    ) public view virtual override returns (bytes4) {
+        bytes1 mode = moduleSignature[0];
+        assembly {
+            let len := mload(moduleSignature)
+            mstore(add(moduleSignature, 0x01), sub(len, 1))
+            moduleSignature := add(moduleSignature, 0x01)
+        }
+
+        // 0x00: signature from spending key
+        // 0x01: aggregated signature from owner and shared secret
+        if (mode == 0x00) {
+            if (_verifySignature(dataHash, moduleSignature, msg.sender)) {
+                return EIP1271_MAGIC_VALUE;
+            }
+            return bytes4(0xffffffff);
+        } else if (mode == 0x01) {
+            if (
+                _verifyAggregateSignature(dataHash, moduleSignature, msg.sender)
+            ) {
+                return EIP1271_MAGIC_VALUE;
+            }
+            return bytes4(0xffffffff);
+        }
+        return bytes4(0xffffffff);
+    }
+
+    /// @inheritdoc ISignatureValidator
+    function isValidSignatureUnsafe(
+        bytes32 dataHash,
+        bytes memory moduleSignature
+    ) public view virtual override returns (bytes4) {
+        bytes1 mode = moduleSignature[0];
+        assembly {
+            let len := mload(moduleSignature)
+            mstore(add(moduleSignature, 0x01), sub(len, 1))
+            moduleSignature := add(moduleSignature, 0x01)
+        }
+
+        // 0x00: signature from spending key
+        // 0x01: aggregated signature from owner and shared secret
+        if (mode == 0x00) {
+            if (_verifySignature(dataHash, moduleSignature, msg.sender)) {
+                return EIP1271_MAGIC_VALUE;
+            }
+            return bytes4(0xffffffff);
+        } else if (mode == 0x01) {
+            if (
+                _verifyAggregateSignature(dataHash, moduleSignature, msg.sender)
+            ) {
+                return EIP1271_MAGIC_VALUE;
+            }
+            return bytes4(0xffffffff);
+        }
+        return bytes4(0xffffffff);
+    }
+
+    /**
+     * @dev Validates a signature for a message signed by address.
+     * @param dataHash Hash of the data.
+     * @param signature Signature to be validated.
+     * @param smartAccount Smart Account address.
+     * @return isValid if signature is valid, false otherwise.
+     */
+    function _verifySignature(
+        bytes32 dataHash,
+        bytes memory signature,
+        address smartAccount
+    ) public view returns (bool) {
+        address stealthAddress = _smartAccountStealth[smartAccount]
+            .stealthAddress;
+        bytes32 hash = ECDSA.toEthSignedMessageHash(dataHash);
+        if (stealthAddress == dataHash.recover(signature)) {
+            return true;
+        }
+        if (stealthAddress != hash.recover(signature)) {
+            return false;
+        }
+        return true;
```

Comment:
Why not
```
if (stealthAddress != hash.recover(signature)) {
            return true;
}
return false;
```
?

Reply:
Can even move
```
bytes32 hash = ECDSA.toEthSignedMessageHash(dataHash);
```
to be after 
```
if (stealthAddress == dataHash.recover(signature)) {}
```
so if `stealthAddress == dataHash.recover(signature)`, the `toEthSignedMessageHash` is not even performed